Split metric reference page routes

kulyk opened the pull request:
Metrics Data Reference page (available at `reference/metrics/:id`) has viewing and edit states. Currently, the page uses the `isEditing` flag in Redux store to switch between these states (via `startEditing` and `endEditing` actions.

This pull request changes the page, so viewing and editing states correspond to different routes (`metrics/:id` for viewing and `metrics/:id/edit` for editing.

kulyk committed:
Add cypress tests for reference/metrics/:id page

kulyk committed:
Add metrics/:metricId/edit route

kulyk committed:
Use router location for metric's isEditing state

Data Reference keeps `isEditing` state flag in Redux store (it's shared across metrics, segments and DB references)

Metrics page now uses router's path name to distinguish reading and editing states

kulyk committed:
Use non-Redux callbacks at MetricDetailContainer

kulyk committed:
Add missing prop type definition

paulrosenzweig commented on frontend/src/metabase/reference/metrics/MetricDetail.jsx:
```diff
@@ -25,14 +25,15 @@ import {
 import * as metadataActions from "metabase/redux/metadata";
 import * as actions from "metabase/reference/reference";
 
+import _ from "underscore";
```
Not a big deal but our general convention is to group external library imports together at the top.

paulrosenzweig commented on frontend/src/metabase/reference/metrics/MetricDetailContainer.jsx:
```diff
@@ -37,19 +32,39 @@ const mapDispatchToProps = {
     await actions.wrappedFetchMetricDetail(this.props, this.props.metricId);
   }
 
+  startEditing() {
```
If you change this to `startEditing = () => {` it's already bound to `this` and you don't need the constructor.

kulyk replied:
Yeah, I wanted to keep following the same approach the rest of the code does. I found a lot of binds in class constructors, so I thought it's the preferred way to bind `this` here. Should we switch to an arrow function here?

paulrosenzweig commented on frontend/src/metabase/reference/metrics/MetricDetailContainer.jsx:
```diff
@@ -37,19 +32,39 @@ const mapDispatchToProps = {
 
+  startEditing() {
+    const { metric, router } = this.props;
+    router.replace(`/reference/metrics/${metric.id}/edit`);
```
`replace` matches the previous behavior, but `push` might be better here. I kinda like being able to go back after clicking "Edit". 

@mazameli or @kdoh do we have a general philosophy on what's a new page? Should switching to edit push a new history item?

paulrosenzweig commented on frontend/src/metabase/reference/metrics/MetricDetailContainer.jsx:
```diff
@@ -63,15 +78,16 @@ export default class MetricDetailContainer extends Component {
   render() {
-    const { isEditing, user, metric } = this.props;
+    const { location, user, metric } = this.props;
+    const isEditing = location.pathname.endsWith("/edit");
```
This creates another place that needs to be kept in sync with `routes.jsx`. If I change the route there from "edit" to "editing", this will break. Instead, maybe pass `isEditing={true}` directly from the route?

e.g. something like this:

```js
<Route
  path="metrics/:metricId/edit"
  render={routeProps =>
    <MetricDetailContainer
      {...routeProps}
      isEditing={true}
    />
  }
/>
```

We still have the push/replace coupled to the route definition, but we should solve that separately by extracting templated route definitions.

kulyk replied:
Good catch, thank you!

kulyk replied:
@paulrosenzweig it appeared to be a bit trickier than I thought

As you can see from the [routes file](https://github.com/metabase/metabase/blob/ea4c591b88af7593508a73b8fa2b4d50d0edc0f9/frontend/src/metabase/routes.jsx#L5), we don't use the default react-routers `Route` component. There is a decorated version of it [here](https://github.com/metabase/metabase/blob/ea4c591b88af7593508a73b8fa2b4d50d0edc0f9/frontend/src/metabase/hoc/Title.jsx#L72) and it basically helps to change pages' meta `title` property.

So, the decorated Route takes a React element passed to it as a `component` prop. Then it wraps it with a HOC that manages meta titles. It does that by wrapping react-router's `Route` static property.

We need to be sure `routeProps` are passed to the final component, but we don't have access to them in Route's static method. Please, correct me if I'm wrong, I can miss something here

I'm worried if we want to make our `Route` support the `render` prop, we'll need to come up with some hacky solution that won't be easy to understand

What do you think if we just import react-router's `Route` here and alias it like `ReactRouterRoute`? It's the simplest thing I see, still, it will be a bit inconsistent and we won't be able to change the meta title for the Metric's reference page (but we don't do it anyway here)

paulrosenzweig replied:
> I'm worried if we want to make our Route support the render prop, we'll need to come up with some hacky solution that won't be easy to understand

That makes sense to me. I think leaving the `isEditing` logic in the component itself is a good solution for now. If we're making similar changes elsewhere maybe we should reconsider and pull it up to the Route.

paulrosenzweig commented on frontend/test/metabase/scenarios/reference/metrics.cy.spec.js:
```diff
@@ -0,0 +1,82 @@
+
+  it("should see the listing", () => {
+    cy.visit("/reference/metrics");
+    cy.contains(METRIC_NAME);
+    cy.contains(METRIC_DESCRIPTION);
```
Ideally, `contains` would be `findByText`. It does an exact match rather than substring match. We use [Cypress testing library](https://testing-library.com/docs/cypress-testing-library/intro/) for this.

paulrosenzweig reviewed (approved):
Looks good!

kulyk committed:
Fix editing callbacks not passed to MetricDetail

kulyk committed:
Fix signInAsAdmin test utility usage

kulyk committed:
Reorder imports

kulyk committed:
Use cy.findByText instead of cy.contains

kulyk committed:
Fix Metrics data reference test

For some reason, when running Cypress tests,
the page appears in `cannot read property fields of undefined`
However it works correctly when running Metabase locally

Checking for undefined `fields` explicitly fixes the problem
Anyway, most likely this component is going to be removed in #15209

kulyk commented:
@paulrosenzweig a few of Metric's data reference cypress tests are failing now because of the #15127 
It has to be fixed by #15209

kulyk committed:
Fix metrics test

paulrosenzweig commented:
The changes from this PR were merged in #15310
